File: ProblemSolve/110.cs
/************************
 * Balanced Binary Tree *
 ************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int HeightOfSubtree(TreeNode root, ref bool ans) {
        if(root == null || ans == false){
            return 0;
        }
        
        int leftSubHeight = HeightOfSubtree(root.left, ref ans);
        int rightSubHeight = HeightOfSubtree(root.right, ref ans);
        
        if((leftSubHeight - rightSubHeight < -1) || (leftSubHeight - rightSubHeight > 1)){
            ans = false;
        }
        
        return 1 + Math.Max(leftSubHeight, rightSubHeight);
    }
    
    public bool IsBalanced(TreeNode root) {
        bool ans = true;
        HeightOfSubtree(root, ref ans);
        return ans;
    }
}

File: ProblemSolve/110.cpp
/************************
 * Balanced Binary Tree *
 ************************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int heightOfSubtree(TreeNode* root, bool& ans) {
        if(root == NULL || ans == false){
            return 0;
        }
        
        int leftSubHeight = heightOfSubtree(root -> left, ans);
        int rightSubHeight = heightOfSubtree(root -> right, ans);
        
        if(leftSubHeight - rightSubHeight < -1 || leftSubHeight - rightSubHeight > 1){
            ans = false;
        }
        
        return 1 + max(leftSubHeight, rightSubHeight);
    }
    
    bool isBalanced(TreeNode* root) {
        bool ans = true;
        heightOfSubtree(root, ans);
        
        return ans;
    }
};